lib/util/request-batching.ts

```diff
@@ -0,0 +1,139 @@
+import * as errors from 'balena-errors';
+import chunk = require('lodash/chunk');
+import { groupByMap, mergePineOptions } from '.';
+import type {
+	PineOptionsWithSelect,
+	PineSelectableProps,
+	PineTypedResult,
+} from '..';
+
+const CHUNK_SIZE = 50;
+
+export function batchResourceOperationFactory<
+	T extends { id: number; uuid: string },
+>({
+	getAll,
+	NotFoundError,
+	AmbiguousResourceError,
+}: {
+	getAll: (options?: PineOptionsWithSelect<T>) => Promise<T[]>;
+	NotFoundError: new (id: string | number) => Error;
+	AmbiguousResourceError: new (id: string | number) => Error;
+}) {
+	type Item<TOpts> = { id: number } & (TOpts extends PineOptionsWithSelect<T>
+		? PineTypedResult<T, TOpts>
+		: {});
+
+	async function batchResourceOperation<
+		TOpts extends PineOptionsWithSelect<T>,
+	>(options: {
+		uuidOrIdOrIds: string | number | number[];
+		options?: TOpts;
+		fn: (items: Array<Item<TOpts>>) => Promise<void>;
+		groupByNavigationPoperty?: undefined;
+	}): Promise<void>;
+	async function batchResourceOperation<
+		TOpts extends PineOptionsWithSelect<T>,
+	>(options: {
+		uuidOrIdOrIds: string | number | number[];
+		options?: TOpts;
+		fn: (items: Array<Item<TOpts>>, ownerId: number) => Promise<void>;
+		groupByNavigationPoperty: PineSelectableProps<T>;
+	}): Promise<void>;
+	async function batchResourceOperation<
+		TOpts extends PineOptionsWithSelect<T>,
+	>({
+		uuidOrIdOrIds,
+		options,
+		groupByNavigationPoperty,
+		fn,
+	}: {
+		uuidOrIdOrIds: string | number | number[];
+		options?: TOpts;
+		fn: (items: Array<Item<TOpts>>, ownerId?: number) => Promise<void>;
+		groupByNavigationPoperty?: PineSelectableProps<T>;
+	}): Promise<void> {
+		if (
+			Array.isArray(uuidOrIdOrIds) &&
+			(!uuidOrIdOrIds.length ||
+				uuidOrIdOrIds.some((id) => typeof id !== 'number'))
+		) {
+			throw new errors.BalenaInvalidParameterError(
+				'uuidOrIdOrIds',
+				uuidOrIdOrIds,
+			);
+		}
+
+		// create a list of UUIDs or chunks of IDs
+		const chunks: Array<string | number[]> =
+			typeof uuidOrIdOrIds === 'string'
+				? [uuidOrIdOrIds]
+				: chunk(
+						!Array.isArray(uuidOrIdOrIds) ? [uuidOrIdOrIds] : uuidOrIdOrIds,
+						CHUNK_SIZE,
+				  );
+
+		const items: Array<
+			Item<TOpts> &
+				Partial<{
+					[groupByNavigationPoperty: string]: { __id: number };
+				}>
+		> = [];
+		for (const uuidOrIdOrIdsChunk of chunks) {
+			const resourceFilter = Array.isArray(uuidOrIdOrIdsChunk)
+				? {
+						id: { $in: uuidOrIdOrIdsChunk },
+				  }
+				: {
+						uuid: { $startswith: uuidOrIdOrIdsChunk },
+				  };
+			const combinedOptions = mergePineOptions(
+				{
+					$select: [
+						'id',
+						...(groupByNavigationPoperty ? [groupByNavigationPoperty] : []),
+					] as Array<PineSelectableProps<T>>,
+					// @ts-expect-error
```

Comment:
I like to add a little blurb as to why we expect the error as it can be pretty helpful later

Reply:
It's b/c `PineFilter<{ id: number; uuid: number; }>` is not assignable to `PineFilter<T>` :(